client/src/main/java/org/eclipse/hono/client/TenantClient.java

```diff
@@ -0,0 +1,43 @@
+/**
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 1.0 which is available at
+ * https://www.eclipse.org/legal/epl-v10.html
+ *
+ * SPDX-License-Identifier: EPL-1.0
+ */
+
+package org.eclipse.hono.client;
+
+import io.vertx.core.Future;
+import org.eclipse.hono.util.TenantObject;
+
+/**
+ * A client for accessing Hono's Tenant API.
+ * <p>
+ * See Hono's <a href="https://www.eclipse.org/hono/api/tenant-api">
+ * Tenant API specification</a> for a description of the result codes returned.
+ * </p>
+ */
+public interface TenantClient extends RequestResponseClient {
+
+    /**
+     * Gets the details of the tenant that this client was created for.
+     * 
+     * @return A future indicating the result of the operation.
+     *         <p>
+     *         The future will succeed if a response with status 200 has been received from the
+     *         tenant service. The JSON object will then contain values as defined in
+     *         <a href="https://www.eclipse.org/hono/api/tenant-api/#get-tenant">
+     *         Get Tenant</a>.
+     *         <p>
+     *         Otherwise, the future will fail with a {@link ServiceInvocationException} containing
+     *         the (error) status code returned by the service.
+     * @see RequestResponseClient#setRequestTimeout(long)
+     */
+    Future<TenantObject> get();
```

Comment:
It feels *very, very* awkward that a `TenantClient` instance is *per-se* scoped to a single tenant.

I understand that we do not really have a good concept for restricting a client's access to a subset of the existing tenant information. But at least we could include a *tenantId* parameter here and then simply authorize the client's request by comparing the identifier of the tenant that the client has been authenticated in  with the tenant ID it provides in the method call. If we then later find a better way of expressing a client's authorities, we at least have the right method signature in place.

Reply:
This is exactly what I had already implemented. And then decided against it again, since this would introduce the "one tenant could wish to access the information of another tenant" - for which we do not have a concept. So I decided to eliminate this param again (check for identity is trivial, but I just ignored this param and still get the tenant from the AMQP app props).
The param felt a little like "premature optimization" to me.

I understand the awkwardness (and find the discussion of exactly this point very helpful!), but this reflects exactly the truth we have now. And forces users of the tenant API to think about why there is no param at all. And if we have a concept, there might be a second operation.

Anyway, I am partly fine with reintroducing the `Future<TenantObject> get(String tenantId)` signature again - but it is not as clear anymore and does not match a pure multi-tenancy concept IMHO. Would then vote for currently ignoring the param, so clients are not forced to provide the tenantId again, which was already provided during client generation like:

`
client.connect(new ProtonClientOptions())
            .compose(c -> c.getOrCreateTenantClient(TEST_HTTP_TENANT))
            .setHandler(ctx.asyncAssertSuccess(r -> {
                tenantClient = r;
            }));
`
So the client could just access the tenant information at other places in the code by 

`
tenantClient.get(null);
`

or (if they have the tenant avail)

`
tenantClient(TEST_HTTP_TENANT);
`

, but they do not **have** to provide their own tenantId again.

WDYT? Reintroduce the tenant param and ignore it? Or check it for identity and force clients to always provide the same tenant again (maybe use null as option to use the already provided tenantId)?

Reply:
FMPOV the client should not need to specify a tenant ID when it creates the `TenantClient`, i.e.
```
client.connect(new ProtonClientOptions())
  .compose(c -> c.getOrCreateTenantClient())
  .setHandler(ctx.asyncAssertSuccess());
```
It can then do `client.get(tenantId)` for whichever tenant it needs information about. Whether the tenant service actually authorizes the request is another topic.

Reply:
BTW the [Tenant API as it is specified currently](https://www.eclipse.org/hono/api/tenant-api/#preconditions) also does not scope the client's links to a particular tenant ID.